MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/query/FingerDrawActivity.java

```diff
@@ -0,0 +1,376 @@
+  private MapboxMap mapboxMap;
+  private FeatureCollection searchPointFeatureCollection;
+  private GeoJsonSource drawLineSource;
+  private GeoJsonSource fillPolygonSource;
+  private List<Point> freehandDrawLineLayerPointList;
+  private List<List<Point>> polygonList;
+
+  /**
+   * Customize search UI with these booleans
+   */
+  private boolean fillSearchAreaWithPolygonWhileDrawing = true;
+  private boolean fillSearchAreaWithPolygon = true;
+  private boolean closePolygonSearchAreaOnceDrawingIsDone = true;
+  private boolean showSearchDataLocations = true;
+
+  private View.OnTouchListener customOnTouchListener = new View.OnTouchListener() {
+    @Override
+    public boolean onTouch(View view, MotionEvent motionEvent) {
+
+      LatLng latLngTouchCoordinate = mapboxMap.getProjection().fromScreenLocation(
+        new PointF(motionEvent.getX(), motionEvent.getY()));
+
+      Point touchPoint = Point.fromLngLat(latLngTouchCoordinate.getLongitude(), latLngTouchCoordinate.getLatitude());
+
+      if (freehandDrawLineLayerPointList != null) {
+
+        // Draw the line as drawing on the map happens
+        freehandDrawLineLayerPointList.add(touchPoint);
+        drawLineSource = mapboxMap.getStyle().getSourceAs(FREEHAND_DRAW_LINE_LAYER_SOURCE_ID);
+        drawLineSource.setGeoJson(LineString.fromLngLats(freehandDrawLineLayerPointList));
+
+        if (fillSearchAreaWithPolygonWhileDrawing) {
+          drawSearchFillArea();
+        }
+
+        // Take certain actions when the drawing is done
+        if (motionEvent.getAction() == MotionEvent.ACTION_UP) {
+          if (closePolygonSearchAreaOnceDrawingIsDone) {
+            freehandDrawLineLayerPointList.add(freehandDrawLineLayerPointList.get(0));
+          }
+
+          if (!fillSearchAreaWithPolygonWhileDrawing && fillSearchAreaWithPolygon) {
+            drawSearchFillArea();
+          }
+
+          if (closePolygonSearchAreaOnceDrawingIsDone) {
+            FeatureCollection pointsInSearchAreaFeatureCollection =
+              TurfJoins.pointsWithinPolygon(searchPointFeatureCollection,
+                FeatureCollection.fromFeature(Feature.fromGeometry(
+                  Polygon.fromLngLats(polygonList))));
+            if (VISIBLE.equals(mapboxMap.getStyle().getLayer(
+              SEARCH_DATA_SYMBOL_LAYER_ID).getVisibility().getValue())) {
+              Toast.makeText(FingerDrawActivity.this, String.format(
+                getString(R.string.search_result_size),
+                pointsInSearchAreaFeatureCollection.features().size()), Toast.LENGTH_SHORT).show();
+            }
+          }
+          enableMapMovement();
+        }
+      }
+      return true;
+    }
+  };
+
+  private void drawSearchFillArea() {
+    // Create and show a FillLayer polygon where the search area is
+    fillPolygonSource = mapboxMap.getStyle().getSourceAs(FREEHAND_DRAW_FILL_LAYER_SOURCE_ID);
+    polygonList = new ArrayList<>();
+    polygonList.add(freehandDrawLineLayerPointList);
```

Comment:
Fill geometries need to be enclosed in all cases, otherwise, the drawn polygon might render incorrectly if it happens to be on a tile boundry.

Reply:
> Fill geometries need to be enclosed

Not sure what you mean and how the code would need to change....

Reply:
First and the last points of the geometry need to be the same.

Reply:
I do this at https://github.com/mapbox/mapbox-android-demo/pull/983/files#diff-e3b54579502b42722d723bdc38541f57R109

Reply:
Correct, but this code is invoked only when the gesture finishes (`ACTION_UP`). This logic has to be performed for every step of the way when it comes to fill geometries.

Reply:
Ok, refactored and came up with ⬇️ 

https://github.com/mapbox/mapbox-android-demo/blob/96c5dc5899efedc324e43821ab6002227fdf090a/MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/query/FingerDrawQueryActivity.java#L104-L117

Largely based on `SatelliteLandSelectActivity`, which is already in this app:

https://github.com/mapbox/mapbox-android-demo/blob/fba1aee8d46ac22ecd498e2c69f461efa94ab83c/MapboxAndroidDemo/src/main/java/com/mapbox/mapboxandroiddemo/examples/dds/SatelliteLandSelectActivity.java#L157-L166